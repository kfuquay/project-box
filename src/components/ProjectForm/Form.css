.create-project-form {
    padding: 0;
    max-width: 800px;
    margin: 0 auto;
}

#add-input-button {
    margin: 5px;
}

#columns {
    display: grid;
    grid-template-columns: 90% 10%;
}

.input-container {
    display: grid;
    grid-template-columns: 100%;
}

.input-container:not(:last-child) {
    margin-bottom: 20px;
}

.input-container > label {
    padding: 8px;
    font-weight: 300;
    letter-spacing: .09em;
    text-transform: uppercase;
    max-width: 220px;
}

.input-section input,
#columns input {
    padding: 15px;
    border: none;
}

#columns button {
    margin-left: auto;
    padding: 8px 16px;
    border: none;
    background: #333;
    color: #f2f2f2;
    text-transform: uppercase;
    letter-spacing: .09em;
    border-radius: 2px;
}

.input-container > input, 
.input-container > button,
#columns > input,
#columns > button {
    margin: 5px;
}

#button-section {
    flex-direction: column;
}

#button-section > button {
    margin: 5px 0 0 auto;
}

.form-section > input {
    margin: 5px;
    display: block;
}