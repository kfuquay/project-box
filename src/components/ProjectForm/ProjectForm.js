import React, { Component } from "react";
import ProjectContext from "../../Context/ProjectContext";
import "./Form.css";

class ProjectForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      project: {
        materials: [{ name: "" }],
        steps: [{ name: "" }],
        title: "",
        summary: "",
        id: "",
      },
    };
  }

  static contextType = ProjectContext;

  componentDidMount() {
    if (this.props.params.edit === "Y") {
      this.setState({
        project: this.props.params.project.project,
      });
    }
  }

  handleTitleChange = e => {
    this.setState({ title: e.target.value });
  };

  handleSummaryChange = e => {
    this.setState({ summary: e.target.value });
  };

  handleMaterialNameChange = index => e => {
    const newMaterials = this.state.materials.map((material, mindex) => {
      if (index !== mindex) return material;
      return { ...material, name: e.target.value };
    });
    this.setState({ materials: newMaterials });
  };

  handleStepNameChange = index => e => {
    const newSteps = this.state.steps.map((step, sindex) => {
      if (index !== sindex) return step;
      return { ...step, name: e.target.value };
    });
    this.setState({ steps: newSteps });
  };

  handleAddMaterial = () => {
    this.setState({
      materials: this.state.materials.concat([{ name: "" }]),
    });
  };

  handleAddStep = () => {
    this.setState({
      steps: this.state.steps.concat([{ name: "" }]),
    });
  };

  handleRemoveMaterial = index => () => {
    this.setState({
      materials: this.state.materials.filter(
        (material, mindex) => index !== mindex
      ),
    });
  };

  handleRemoveStep = index => () => {
    this.setState({
      steps: this.state.steps.filter((step, sindex) => index !== sindex),
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const newProject = {
      id: Math.random() * 1000,
      title: this.state.title,
      img:
        "https://images.pexels.com/photos/70083/frog-macro-amphibian-green-70083.jpeg?cs=srgb&dl=amphibian-color-colour-70083.jpg&fm=jpg",
      summary: this.state.summary,
      materials: this.state.materials,
      steps: this.state.steps,
    };
    this.context.handleSubmitNewProject(newProject);
  };

  render() {
    return (
      <section className="form-container">
        <form className="create-project-form" onSubmit={this.handleSubmit}>
          <div className="input-container">
            <label htmlFor="title">Project title</label>
            <input
              type="text"
              name="title"
              placeholder="knit sweater"
              value={this.state.project.title}
              onChange={this.handleTitleChange}
              required
            />
          </div>
          <div className="input-container">
            <label htmlFor="summary">Summary</label>
            <textarea
              name="summary"
              rows="15"
              value={this.state.project.summary}
              onChange={this.handleSummaryChange}
            />
          </div>
          <div className="input-container">
            <label htmlFor="material">Materials</label>
            {this.state.project.materials.map((material, index) => (
              <div id="columns">
                <input
                  type="text"
                  name="material"
                  value={material.name}
                  onChange={this.handleMaterialNameChange(index)}
                />
                <button
                  type="button"
                  onClick={this.handleRemoveMaterial(index)}
                >
                  {" "}
                  -{" "}
                </button>
              </div>
            ))}
          </div>
          <div className="form-section button-section">
            <button type="button" onClick={this.handleAddMaterial}>
              + Material
            </button>
          </div>

          <div className="input-container">
            <label htmlFor="step">Steps</label>
            {this.state.project.steps.map((step, index) => (
              <div id="columns">
                <input
                  type="text"
                  name="step"
                  value={step.name}
                  onChange={this.handleStepNameChange(index)}
                />
                <button type="button" onClick={this.handleRemoveStep(index)}>
                  {" "}
                  -{" "}
                </button>
              </div>
            ))}
          </div>
          <div className="form-section button-section">
            <button type="button" onClick={this.handleAddStep}>
              + Step
            </button>
          </div>

          <div className="form-section" id="button-section">
            <button type="submit">Submit</button>
          </div>
        </form>
      </section>
    );
  }
}

export default ProjectForm;
