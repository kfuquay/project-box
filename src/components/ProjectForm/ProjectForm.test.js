import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import ProjectForm from "./ProjectForm";

describe(`ProjectForm  component`, () => {
  it.skip("renders without crashing", () => {
    const extras = {
      params: {
        props: {
          edit: "N",
        },
      },
    };
    const wrapper = shallow(<ProjectForm {...extras} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
