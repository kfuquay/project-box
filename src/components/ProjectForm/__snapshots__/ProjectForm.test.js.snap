// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm  component renders without crashing 1`] = `
<section
  className="form-container"
>
  <form
    className="create-project-form"
    onSubmit={[Function]}
  >
    <div
      className="input-container"
    >
      <label
        htmlFor="title"
      >
        Project title
      </label>
      <input
        aria-required="true"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="knit sweater"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-container"
    >
      <label
        htmlFor="summary"
      >
        Summary
      </label>
      <textarea
        id="summary"
        name="summary"
        onChange={[Function]}
        rows="15"
        value=""
      />
    </div>
    <div
      className="input-container"
    >
      <label
        htmlFor="material"
      >
        Materials
      </label>
      <div
        id="columns"
        key="0"
      >
        <input
          aria-labelledby="material"
          name="material"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          onClick={[Function]}
          type="button"
        >
           
          -
           
        </button>
      </div>
    </div>
    <div
      className="form-section button-section"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        + Material
      </button>
    </div>
    <div
      className="input-container"
    >
      <label
        htmlFor="step"
      >
        Steps
      </label>
      <div
        id="columns"
        key="0"
      >
        <input
          aria-labelledby="step"
          name="step"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          onClick={[Function]}
          type="button"
        >
           
          -
           
        </button>
      </div>
    </div>
    <div
      className="form-section button-section"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        + Step
      </button>
    </div>
    <div
      className="form-section"
      id="button-section"
    >
      <button
        type="submit"
      >
        Submit
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</section>
`;
